chameleon-core/src/main/java/org/hibernate/omm/jdbc/MongoPreparedStatement.java

```diff
@@ -100,149 +106,112 @@ public void addBatch() {
     }
 
     @Override
-    public void setTime(final int parameterIndex, @Nullable final Time x) {
-        parameters.put(parameterIndex, x == null ? "null" : "{\"$date\": {\"$numberLong\": \"" + x.toInstant().getEpochSecond() + "\"}}");
+    public void setTime(final int parameterIndex, final Time x) {
+        parameters.put(parameterIndex, new BsonDateTime(x.toInstant().toEpochMilli()));
     }
 
     @Override
-    public void setTimestamp(final int parameterIndex, @Nullable final Timestamp x) {
-        parameters.put(
-                parameterIndex,
-                x == null ? "null" : "{\"$timestamp\": {\"" + x.toInstant().getEpochSecond() + "\": <t>, \"i\": 1}}"
+    public void setTimestamp(final int parameterIndex, final Timestamp x) {
+        parameters.put(parameterIndex, new BsonDateTime(x.toInstant().toEpochMilli())
+
         );
     }
 
     @Override
-    public void setObject(final int parameterIndex, @Nullable final Object x, final int targetSqlType)
+    public void setObject(final int parameterIndex, final Object x, final int targetSqlType)
             throws SimulatedSQLException {
-        if (x == null) {
-            parameters.put(parameterIndex, "null");
-        } else {
-            switch (targetSqlType) {
-                case 3_000:
-                    ObjectId objectId = (ObjectId) x;
-                    parameters.put(parameterIndex, "{ \"$oid\": \"" + objectId.toHexString() + "\" }");
-                    break;
-                default:
-                    throw new NotSupportedSQLException("unknown MongoSqlType: " + targetSqlType);
-            }
+        switch (targetSqlType) {
+            case 3_000:
+                parameters.put(parameterIndex, new BsonObjectId((ObjectId) x));
+                break;
+            default:
+                throw new NotSupportedSQLException("unknown MongoSqlType: " + targetSqlType);
         }
     }
 
     @Override
     public void setArray(final int parameterIndex, final Array x) throws SimulatedSQLException {
-        Assertions.notNull("x", x);
-
         try {
-            Iterable<?> iterable;
-            if (x.getArray().getClass().isArray()) {
-                iterable = Arrays.asList((Object[]) x.getArray());
-            } else {
-                iterable = ((Iterable<?>) x.getArray());
+            Object[] objectArray = (Object[]) x.getArray();
+            BsonArray bsonArray = new BsonArray(objectArray.length);
+            for (int i = 0; i < objectArray.length; i++) {
+                Object objectValue = objectArray[i];
+                BsonValue bsonValue;
+                if (objectValue instanceof String) {
+                    bsonValue = new BsonString((String) objectValue);
+                } else if (objectValue instanceof Integer) {
+                    bsonValue = new BsonInt32((Integer) objectValue);
+                } else {
+                    // TODO: support the rest of the types
```

Comment:
There are only tests for strings so didn't bother to add any more value types here yet.

Reply:
yeah, this repo is mainly for POC purpose, not for prod. All are good.